website/mailing_list/utils.py

```diff
@@ -0,0 +1,422 @@
+# Signalled Functions
+###############################################################################
+
+@contributor_added.connect
+def subscribe_contributor_to_mailing_list(node, contributor, auth=None):
+    if node.mailing_enabled and contributor.is_active:
+        subscription = node.get_or_create_mailing_list_subscription()
+        subscription.add_user_to_subscription(contributor, 'email_transactional', save=True)
+        celery_update_single_user_in_list(node._id, contributor._id)
+        node.mailing_updated = True
+        node.save()
+
+@contributor_removed.connect
+def unsubscribe_contributor_from_mailing_list(node, user, auth=None):
+    if node.mailing_enabled:
+        subscription = node.get_or_create_mailing_list_subscription()
+    else:
+        from website.models import NotificationSubscription  # avoid circular import
+        subscription = NotificationSubscription.load(to_subscription_key(node._id, 'mailing_list_events'))
+    if subscription:
+        subscription.remove_user_from_subscription(user)
+        node.mailing_updated = True
+        node.save()
+    celery_remove_user_from_list(node._id, user._id)
+
+
+@user_confirmed.connect
+def resubscribe_on_confirm(user):
+    for node in user.contributed:
+        subscribe_contributor_to_mailing_list(node, user)
+
+@node_deleted.connect
+def remove_list_for_deleted_node(node):
+    node.mailing_updated = True
+    node.save()
+    celery_delete_list(node._id)
+
+
+###############################################################################
+# Mailing List Helper Functions
+###############################################################################
+
+def jsonify_users_list(users, unsubs=[]):
+    """ Serializes list of users for Mailgun
+
+    :param list users: users to serialize
+    :param list unsubs: unsubscribed users
+    """
+    members_list = []
+    for member in users:
+        for email in member.emails:
+            members_list.append({
+                'address': email,
+                'subscribed': member not in unsubs and email == member.username,
+                'vars': {'_id': member._id, 'primary': email == member.username}
+            })
+    return members_list
+
+def address(node_id):
+    return '{}@{}'.format(node_id, furl(settings.DOMAIN).host)
+
+def list_title(node):
+    return '{} Mailing List'.format(node.title)
+
+def find_email(long_email):
+    # allow for both "{email}" syntax and "{name} <{email}>" syntax
+    if '<' in long_email:
+        email_match = ANGLE_BRACKETS_REGEX.search(long_email)
+        if email_match:
+            return email_match.groups()[0].lower().strip()
+    elif '@' in long_email:
+        return long_email.lower().strip()
+    return None
+
+def log_message(**kwargs):
+    """ Acquires and logs messages sent through Mailgun"""
+    from website.models import Node  # avoid circular imports
+    message = request.form
```

Comment:
By picking the message off from the `flask.request.form` parameter (rather than acting on data passed in as an argument), you couple this function tightly to a running flask application.

In the future we may want all work to be done by an API (eg Django). When possible, I recommend writing all utility functions so that they act on data passed explicitly in- this will make it easier to decouple the code or use it in a different backend later.